lib/ansible/plugins/connections/ssh.py

```diff
@@ -63,38 +57,79 @@ def set_host_overrides(self, host):
         if 'ansible_ssh_args' in v:
             self.ssh_args = v['ansible_ssh_args']
 
-    @property
-    def transport(self):
-        ''' used to identify this connection object from other classes '''
-        return 'ssh'
+    # The connection is created by running ssh/scp/sftp from the exec_command,
+    # put_file, and fetch_file methods, so we don't need to do any connection
+    # management here.
 
-    def _split_args(self, argstring):
-        """
-        Takes a string like '-o Foo=1 -o Bar="foo bar"' and returns a
-        list ['-o', 'Foo=1', '-o', 'Bar=foo bar'] that can be added to
-        the argument list. The list will not contain any empty elements.
-        """
-        return [x.strip() for x in shlex.split(argstring) if x.strip()]
+    def _connect(self):
+        self._connected = True
+        return self
 
-    def add_args(self, explanation, args):
-        """
-        Adds the given args to _common_args and displays a
-        caller-supplied explanation of why they were added.
-        """
-        self._common_args += args
-        self._display.vvvvv('SSH: ' + explanation + ': (%s)' % ')('.join(args), host=self._play_context.remote_addr)
+    def close(self):
+        # If we have a persistent ssh connection (ControlPersist), we can ask it
+        # to stop listening. Otherwise, there's nothing to do here.
 
-    def _connect(self):
-        ''' connect to the remote host '''
+        if self._connected and self._persistent:
+            cmd = self._build_command('ssh', '-O', 'stop', self.host)
+            p = subprocess.Popen(cmd, stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+            stdout, stderr = p.communicate()
 
-        self._display.vvv("ESTABLISH SSH CONNECTION FOR USER: {0}".format(self._play_context.remote_user), host=self._play_context.remote_addr)
+        self._connected = False
+
+    def _build_command(self, binary, *other_args):
+        '''
+        Takes a binary (ssh, scp, sftp) and optional extra arguments and returns
+        a command line as an array that can be passed to subprocess.Popen after
+        appending any extra commands to it.
```

Comment:
Well, I was thinking of "ssh <options> <hostname> 'some command with sudo and sh and whatnot'". Arguments would work too. I'll change it.